File: apps/web/src/agents/accountantAgent.ts
import { Type } from "@google/genai";
import type { AnalysisResult, AuditReport, AccountingEntry, AuditedDocument, SpedFile } from '../types';
import { logger } from "../services/logger";
import { parseSafeFloat } from "../utils/parsingUtils";
import { generateJSON } from "../services/geminiService";

const analysisResponseSchema = {
  type: Type.OBJECT,
  properties: {
    title: { type: Type.STRING },
    summary: { type: Type.STRING },
    keyMetrics: {
      type: Type.ARRAY,
      items: {
        type: Type.OBJECT,
        properties: {
          metric: { type: Type.STRING },
          value: { type: Type.STRING },
          insight: { type: Type.STRING },
        },
        required: ['metric', 'value', 'insight'],
      },
    },
    actionableInsights: {
      type: Type.ARRAY,
      items: { type: Type.STRING },
    },
    strategicRecommendations: {
        type: Type.ARRAY,
        items: { type: Type.STRING },
        description: 'Recomendações estratégicas de alto nível para o negócio.'
    }
  },
  required: ['title', 'summary', 'keyMetrics', 'actionableInsights', 'strategicRecommendations'],
};

const formatCurrency = (value: number) => value.toLocaleString('pt-BR', { style: 'currency', currency: 'BRL' });

const MOCK_ALIQUOTAS = {
    IVA: 0.25, // 25% (simulado)
};

/**
 * Rebuilt deterministic aggregation logic inspired by the "TotalSynth Core" manifest.
 * It is now the single source of truth for calculating aggregated metrics, ensuring consistency and robustness.
 * @param report The audit report containing all processed documents.
 * @returns A record of aggregated metrics.
 */
const runDeterministicAccounting = (report: Omit<AuditReport, 'summary'>): Record<string, any> => {
    const validDocs = report.documents.filter(d => d.status !== 'ERRO' && d.doc.data && d.doc.data.length > 0);

    const defaultMetrics = {
        'Número de Documentos Válidos': 0,
        'Valor Total das NFes': formatCurrency(0),
        'Valor Total dos Produtos': formatCurrency(0),
        'Total de Itens Processados': 0,
        'Valor Total de ICMS': formatCurrency(0),
        'Valor Total de PIS': formatCurrency(0),
        'Valor Total de COFINS': formatCurrency(0),
        'Valor Total de ISS': formatCurrency(0),
        'Estimativa de IVA (Simulado)': formatCurrency(0),
    };

    if (validDocs.length === 0) {
        return defaultMetrics;
    }

    const allItems = validDocs.flatMap(d => d.doc.data!);
    
    // [CRITICAL FIX] Use a Map to correctly sum the total values of unique NFes and count them.
    // This resolves the "0 Documentos Válidos" bug by relying on the now-correctly-extracted `nfe_id`.
    const uniqueNfes = new Map<string, { totalValue: number }>();
    for (const item of allItems) {
        if (item.nfe_id) {
            // The `valor_total_nfe` is repeated for each item, so we can just set it.
            // The Map ensures we only store the value once per unique NFe ID.
            uniqueNfes.set(item.nfe_id, { totalValue: parseSafeFloat(item.valor_total_nfe) });
        }
    }
    
    let totalNfeValue = Array.from(uniqueNfes.values()).reduce((sum, nfe) => sum + nfe.totalValue, 0);
    
    const totals = allItems.reduce((acc, item) => {
        acc.totalProductValue += parseSafeFloat(item.produto_valor_total);
        acc.totalICMS += parseSafeFloat(item.produto_valor_icms);
        acc.totalPIS += parseSafeFloat(item.produto_valor_pis);
        acc.totalCOFINS += parseSafeFloat(item.produto_valor_cofins);
        acc.totalISS += parseSafeFloat(item.produto_valor_iss);
        return acc;
    }, { totalProductValue: 0, totalICMS: 0, totalPIS: 0, totalCOFINS: 0, totalISS: 0 });

    // --- [TotalSynth] Fallback Logic ---
    // If the official total NFe value is still zero but there are product values,
    // reconstruct the total as a fail-safe.
    if (totalNfeValue === 0 && totals.totalProductValue > 0) {
        logger.log('AccountantAgent', 'WARN', '[TotalSynth] Valor total da NFe é zero. Reconstruindo a partir da soma dos componentes como fallback.');
        totalNfeValue = totals.totalProductValue + totals.totalICMS + totals.totalPIS + totals.totalCOFINS + totals.totalISS;
    }

    const totalIVA = (totals.totalPIS + totals.totalCOFINS) * MOCK_ALIQUOTAS.IVA;

    const metrics: Record<string, string | number> = {
        'Número de Documentos Válidos': uniqueNfes.size, // Correctly counts unique NFes
        'Valor Total das NFes': formatCurrency(totalNfeValue),
        'Valor Total dos Produtos': formatCurrency(totals.totalProductValue),
        'Total de Itens Processados': allItems.length,
        'Valor Total de ICMS': formatCurrency(totals.totalICMS),
        'Valor Total de PIS': formatCurrency(totals.totalPIS),
        'Valor Total de COFINS': formatCurrency(totals.totalCOFINS),
        'Valor Total de ISS': formatCurrency(totals.totalISS),
        'Estimativa de IVA (Simulado)': formatCurrency(totalIVA),
    };
    
    if (totalNfeValue === 0 && allItems.length > 0) {
        metrics['Alerta de Qualidade'] = 'O valor total das NFes processadas é zero, mesmo após as tentativas de reconstrução. Isso indica problemas graves nos dados de origem.';
    }

    return metrics;
}

const runAIAccountingSummary = async (dataSample: string, aggregatedMetrics: Record<string, any>): Promise<AnalysisResult> => {
  const dataQualityIssue = aggregatedMetrics['Alerta de Qualidade'];

  const prompt = `
        You are an expert financial analyst. I have performed a preliminary, deterministic analysis on a batch of fiscal documents and derived the following key aggregated metrics:
        ---
        Aggregated Metrics:
        ${JSON.stringify(aggregatedMetrics, null, 2)}
        ---
        ${dataQualityIssue ? `\n**ALERTA CRÍTICO DE QUALIDADE DE DADOS:** ${dataQualityIssue}\n` : ''}
        I also have a small, representative sample of the line-item data from these documents in CSV format:
        ---
        Data Sample:
        ${dataSample}
        ---

        Your task is to act as the final step in the analysis pipeline.
        1.  Create a compelling, professional 'title' for this analysis report.
        2.  Write a concise 'summary' of the fiscal situation based on both the aggregated metrics and the data sample.
        3.  Populate the 'keyMetrics' array. You MUST use the pre-calculated aggregated metrics as the primary source of truth. You can add 1-2 additional metrics if you can derive them reliably from the data sample (e.g., Top Product by Value).
        4.  Generate 2-3 insightful, 'actionableInsights' for a business manager. ${dataQualityIssue ? "Você DEVE incluir um insight abordando diretamente o alerta de qualidade de dados." : ""}
        5.  Based on everything, provide 1-2 'strategicRecommendations' for the business. These should be higher-level than the actionable insights, focusing on long-term strategy (e.g., "Consider reviewing the supply chain for product X due to price volatility," or "Evaluate the tax regime for service-based revenue.").

        The entire response must be in Brazilian Portuguese and formatted as a single JSON object adhering to the required schema. Do not include any text outside of the JSON object.
    `;
  
  return generateJSON<AnalysisResult>(
    'gemini-2.5-flash',
    prompt,
    analysisResponseSchema
  );
};

const generateAccountingEntries = (documents: AuditedDocument[]): AccountingEntry[] => {
    const entries: AccountingEntry[] = [];
    
    for (const doc of documents) {
        if (doc.status === 'ERRO' || !doc.classification || !doc.doc.data || doc.doc.data.length === 0) continue;
        
        const totalNfe = parseSafeFloat(doc.doc.data[0]?.valor_total_nfe);
        const totalProducts = doc.doc.data.reduce((sum, item) => sum + parseSafeFloat(item.produto_valor_total), 0);
        const totalIcms = doc.doc.data.reduce((sum, item) => sum + parseSafeFloat(item.produto_valor_icms), 0);
        
        if (totalNfe === 0 && totalProducts === 0) continue; // Skip docs with no values

        switch (doc.classification.operationType) {
            case 'Compra':
                entries.push({ docName: doc.doc.name, account: '1.1.2 Estoques', type: 'D', value: totalProducts });
                if (totalIcms > 0) {
                    entries.push({ docName: doc.doc.name, account: '1.2.1 ICMS a Recuperar', type: 'D', value: totalIcms });
                }
                entries.push({ docName: doc.doc.name, account: '2.1.1 Fornecedores', type: 'C', value: totalNfe });
                break;
            case 'Venda':
                entries.push({ docName: doc.doc.name, account: '1.1.3 Clientes', type: 'D', value: totalNfe });
                entries.push({ docName: doc.doc.name, account: '4.1.1 Receita de Vendas', type: 'C', value: totalProducts });
                if (totalIcms > 0) {
                     entries.push({ docName: doc.doc.name, account: '4.2.1 ICMS sobre Vendas', type: 'D', value: totalIcms });
                     entries.push({ docName: doc.doc.name, account: '2.1.2 ICMS a Recolher', type: 'C', value: totalIcms });
                }
                break;
            case 'Devolução':
                const firstCfopDev = doc.doc.data[0]?.produto_cfop?.toString();
                if (firstCfopDev?.startsWith('1') || firstCfopDev?.startsWith('2')) { // Devolução de Compra
                    entries.push({ docName: doc.doc.name, account: '2.1.1 Fornecedores', type: 'D', value: totalNfe });
                    if (totalIcms > 0) {
                        entries.push({ docName: doc.doc.name, account: '1.2.1 ICMS a Recuperar', type: 'C', value: totalIcms });
                    }
                    entries.push({ docName: doc.doc.name, account: '1.1.2 Estoques', type: 'C', value: totalProducts });
                } else if (firstCfopDev?.startsWith('5') || firstCfopDev?.startsWith('6')) { // Devolução de Venda
                    entries.push({ docName: doc.doc.name, account: '4.1.2 Devoluções de Vendas', type: 'D', value: totalProducts });
                    if (totalIcms > 0) {
                        entries.push({ docName: doc.doc.name, account: '2.1.2 ICMS a Recolher', type: 'D', value: totalIcms });
                    }
                    entries.push({ docName: doc.doc.name, account: '1.1.3 Clientes', type: 'C', value: totalNfe });
                }
                break;
            case 'Serviço':
                const firstCfopServ = doc.doc.data[0]?.produto_cfop?.toString();
                 if (firstCfopServ?.startsWith('5') || firstCfopServ?.startsWith('6') || firstCfopServ?.startsWith('7')) { // Serviço Prestado
                    entries.push({ docName: doc.doc.name, account: '1.1.3 Clientes', type: 'D', value: totalNfe });
                    entries.push({ docName: doc.doc.name, account: '4.1.3 Receita de Serviços', type: 'C', value: totalNfe });
                } else { // Serviço Tomado (compra)
                    entries.push({ docName: doc.doc.name, account: '3.1.1 Despesa com Serviços', type: 'D', value: totalNfe });
                    entries.push({ docName: doc.doc.name, account: '2.1.1 Fornecedores', type: 'C', value: totalNfe });
                }
                break;
            case 'Transferência':
                const firstCfopTransf = doc.doc.data[0]?.produto_cfop?.toString();
                if (firstCfopTransf?.startsWith('5') || firstCfopTransf?.startsWith('6')) { // Transferência de saída
                    entries.push({ docName: doc.doc.name, account: '3.1.2 Custo de Transferência', type: 'D', value: totalProducts });
                    entries.push({ docName: doc.doc.name, account: '1.1.2 Estoques', type: 'C', value: totalProducts });
                } else { // Transferência de entrada
                    entries.push({ docName: doc.doc.name, account: '1.1.2 Estoques', type: 'D', value: totalProducts });
                    entries.push({ docName: doc.doc.name, account: '4.1.4 Receita de Transferência', type: 'C', value: totalProducts });
                }
                break;
        }
    }
    return entries;
};

const generateSpedEfd = (report: Pick<AuditReport, 'documents'>): SpedFile => {
    const lines: string[] = [];
    const today = new Date();
    const dataIni = new Date(today.getFullYear(), today.getMonth(), 1).toLocaleDateString('pt-BR').replace(/\//g, '');
    const dataFim = new Date(today.getFullYear(), today.getMonth() + 1, 0).toLocaleDateString('pt-BR').replace(/\//g, '');

    const recordCounts: Record<string, number> = {};
    const countRecord = (type: string) => { recordCounts[type] = (recordCounts[type] || 0) + 1; };

    // Bloco 0
    lines.push(`|0000|017|0|${dataIni}|${dataFim}|Nexus QuantumI2A2|12345678000195||SP|||A|1|`);
    countRecord('0000');
    lines.push('|0001|0|');
    countRecord('0001');

    // Bloco C
    lines.push('|C001|0|');
    countRecord('C001');
    const validDocs = report.documents.filter(d => d.status !== 'ERRO' && d.doc.data && d.doc.data.length > 0);
    
    for(const doc of validDocs) {
        const firstItem = doc.doc.data![0];
        lines.push(`|C100|${doc.classification?.operationType === 'Compra' ? '0' : '1'}|0||55|||${parseSafeFloat(firstItem.valor_total_nfe).toFixed(2).replace('.',',')}||||||||`);
        countRecord('C100');

        const c190Aggregator: Record<string, { vBC: number, vIcms: number, vOper: number }> = {};
        
        doc.doc.data?.forEach((item) => {
            const cst = (item.produto_cst_icms?.toString() || '00').slice(0, 3);
            const cfop = item.produto_cfop?.toString() || '0000';
            const aliq = (item.produto_aliquota_icms || 0).toFixed(2).replace('.',',');
            const key = `${cst}|${cfop}|${aliq}`;

            if (!c190Aggregator[key]) {
                c190Aggregator[key] = { vBC: 0, vIcms: 0, vOper: 0 };
            }
            c190Aggregator[key].vBC += parseSafeFloat(item.produto_base_calculo_icms);
            c190Aggregator[key].vIcms += parseSafeFloat(item.produto_valor_icms);
            c190Aggregator[key].vOper += parseSafeFloat(item.produto_valor_total);
        });

        Object.entries(c190Aggregator).forEach(([key, values]) => {
            const [cst, cfop, aliq] = key.split('|');
            lines.push(`|C190|${cst}|${cfop}|${aliq}|${values.vOper.toFixed(2).replace('.',',')}|${values.vBC.toFixed(2).replace('.',',')}|${values.vIcms.toFixed(2).replace('.',',')}||||`);
            countRecord('C190');
        });
        
        doc.doc.data?.forEach((item, index) => {
            lines.push(`|C170|${index+1}|${item.produto_nome || ''}|${parseSafeFloat(item.produto_qtd).toFixed(2).replace('.',',')}|UN|${parseSafeFloat(item.produto_valor_total).toFixed(2).replace('.',',')}||${item.produto_cfop}|${item.produto_cst_icms}||||`);
            countRecord('C170');
        });
    }
    
    lines.push(`|C990|${1 + (recordCounts['C100'] || 0) + (recordCounts['C170'] || 0) + (recordCounts['C190'] || 0) + 1}|`);
    countRecord('C990');
    
    // Bloco 9
    lines.push('|9001|0|');
    countRecord('9001');

    lines.push('|0990|2|');
    countRecord('0990');
    
    const finalRecordCounts = { ...recordCounts };
    finalRecordCounts['9990'] = 1; // Self-reference for 9990
    finalRecordCounts['9999'] = 1; // Self-reference for 9999
    
    const sortedRecords = Object.keys(finalRecordCounts).sort();
    sortedRecords.forEach(rec => {
        lines.push(`|9900|${rec}|${finalRecordCounts[rec]}|`);
        countRecord('9900');
    });

    lines.push(`|9990|${(recordCounts['9001'] || 0) + (recordCounts['9900'] || 0) + 1}|`);
    countRecord('9990');

    const totalLines = lines.length + 1;
    lines.push(`|9999|${totalLines}|`);

    return {
        filename: `SPED-EFD-${today.toISOString().split('T')[0]}.txt`,
        content: lines.join('\n')
    };
};

export const runAccountingAnalysis = async (report: Omit<AuditReport, 'summary'>): Promise<AuditReport> => {
    // 1. Run deterministic calculations first
    const aggregatedMetrics = runDeterministicAccounting(report);

    // 2. Generate Accounting Entries
    const accountingEntries = generateAccountingEntries(report.documents);
    
    // 3. Generate SPED File
    const spedFile = generateSpedEfd(report);

    const validDocsData = report.documents
        .filter(d => d.status !== 'ERRO' && d.doc.data)
        .flatMap(d => d.doc.data!);
    
    if (validDocsData.length === 0) {
        // Return a default summary if no valid data is available
        const defaultSummary: AnalysisResult = {
            title: "Análise Fiscal Concluída",
            summary: "Não foram encontrados dados válidos para gerar um resumo detalhado. Verifique os documentos com erro.",
            keyMetrics: Object.entries(aggregatedMetrics).map(([key, value]) => ({ metric: key, value: String(value), insight: "" })),
            actionableInsights: ["Verificar a causa dos erros nos documentos importados para permitir uma análise completa."],
            strategicRecommendations: ["Implementar um processo de validação de arquivos na origem para garantir a qualidade dos dados para análise."]
        };
         return { ...report, summary: defaultSummary, aggregatedMetrics, accountingEntries, spedFile };
    }
    
    const { default: Papa } = await import('papaparse');
    const dataSampleForAI = Papa.unparse(validDocsData.slice(0, 200));

    // 4. Run AI analysis with deterministic data as context
    const summary = await runAIAccountingSummary(dataSampleForAI, aggregatedMetrics);

    // 5. Combine results
    return { ...report, summary, aggregatedMetrics, accountingEntries, spedFile };
};

File: backend/agents/accountant_agent.py
from __future__ import annotations

import unicodedata
from collections import defaultdict
from datetime import date
from typing import Dict, List

from backend.types import (
    AccountingEntry,
    AnalysisResult,
    AuditReport,
    AuditedDocument,
    SpedFile,
)
from backend.utils.parsing import parse_safe_float


def _format_currency(value: float) -> str:
    return f"R$ {value:,.2f}".replace(",", "X").replace(".", ",").replace("X", ".")


MOCK_ALIQUOTAS = {"IVA": 0.25}


_METRIC_LABELS = {
    "documents": "N\u00famero de Documentos V\u00e1lidos",
    "nfes": "Valor Total das NFes",
    "products": "Valor Total dos Produtos",
    "items": "Total de Itens Processados",
    "icms": "Valor Total de ICMS",
    "pis": "Valor Total de PIS",
    "cofins": "Valor Total de COFINS",
    "iss": "Valor Total de ISS",
    "iva": "Estimativa de IVA (Simulado)",
}

_METRIC_KEY_CANONICAL = {
    "Numero de Documentos Validos": _METRIC_LABELS["documents"],
    "Ngmero de Documentos Volidos": _METRIC_LABELS["documents"],
    _METRIC_LABELS["documents"]: _METRIC_LABELS["documents"],
    _METRIC_LABELS["nfes"]: _METRIC_LABELS["nfes"],
    _METRIC_LABELS["products"]: _METRIC_LABELS["products"],
    _METRIC_LABELS["items"]: _METRIC_LABELS["items"],
    _METRIC_LABELS["icms"]: _METRIC_LABELS["icms"],
    _METRIC_LABELS["pis"]: _METRIC_LABELS["pis"],
    _METRIC_LABELS["cofins"]: _METRIC_LABELS["cofins"],
    _METRIC_LABELS["iss"]: _METRIC_LABELS["iss"],
    _METRIC_LABELS["iva"]: _METRIC_LABELS["iva"],
}


def _normalize_metric_keys(metrics: Dict[str, object]) -> Dict[str, object]:
    normalized: Dict[str, object] = {}
    for key, value in metrics.items():
        ascii_key = unicodedata.normalize("NFKD", key).encode("ascii", "ignore").decode("ascii")
        canonical = _METRIC_KEY_CANONICAL.get(ascii_key, _METRIC_KEY_CANONICAL.get(key, key))
        normalized[canonical] = value
    return normalized


def _run_deterministic_accounting(report: AuditReport) -> Dict[str, object]:
    valid_docs = [doc for doc in report.documents if doc.status != doc.status.ERRO and doc.doc.data]
    metrics = {
        _METRIC_LABELS["documents"]: 0,
        _METRIC_LABELS["nfes"]: _format_currency(0),
        _METRIC_LABELS["products"]: _format_currency(0),
        _METRIC_LABELS["items"]: 0,
        _METRIC_LABELS["icms"]: _format_currency(0),
        _METRIC_LABELS["pis"]: _format_currency(0),
        _METRIC_LABELS["cofins"]: _format_currency(0),
        _METRIC_LABELS["iss"]: _format_currency(0),
        _METRIC_LABELS["iva"]: _format_currency(0),
    }
    if not valid_docs:
        return _normalize_metric_keys(metrics)

    all_items = [item for doc in valid_docs for item in (doc.doc.data or [])]
    unique_nfes: Dict[str, float] = {}
    for item in all_items:
        nfe_id = item.get("nfe_id")
        if not nfe_id:
            continue
        unique_nfes[str(nfe_id)] = parse_safe_float(item.get("valor_total_nfe"))

    totals = defaultdict(float)
    for item in all_items:
        totals["totalProductValue"] += parse_safe_float(item.get("produto_valor_total"))
        totals["totalICMS"] += parse_safe_float(item.get("produto_valor_icms"))
        totals["totalPIS"] += parse_safe_float(item.get("produto_valor_pis"))
        totals["totalCOFINS"] += parse_safe_float(item.get("produto_valor_cofins"))
        totals["totalISS"] += parse_safe_float(item.get("produto_valor_iss"))

    total_nfe_value = sum(unique_nfes.values())
    if total_nfe_value == 0 and totals["totalProductValue"] > 0:
        total_nfe_value = (
            totals["totalProductValue"]
            + totals["totalICMS"]
            + totals["totalPIS"]
            + totals["totalCOFINS"]
            + totals["totalISS"]
        )

    total_iva = (totals["totalPIS"] + totals["totalCOFINS"]) * MOCK_ALIQUOTAS["IVA"]

    metrics.update(
        {
            _METRIC_LABELS["documents"]: len(unique_nfes) or len(valid_docs),
            _METRIC_LABELS["nfes"]: _format_currency(total_nfe_value),
            _METRIC_LABELS["products"]: _format_currency(totals["totalProductValue"]),
            _METRIC_LABELS["items"]: len(all_items),
            _METRIC_LABELS["icms"]: _format_currency(totals["totalICMS"]),
            _METRIC_LABELS["pis"]: _format_currency(totals["totalPIS"]),
            _METRIC_LABELS["cofins"]: _format_currency(totals["totalCOFINS"]),
            _METRIC_LABELS["iss"]: _format_currency(totals["totalISS"]),
            _METRIC_LABELS["iva"]: _format_currency(total_iva),
        }
    )

    if total_nfe_value == 0 and all_items:
        metrics["Alerta de Qualidade"] = (
            "O valor total das NFes processadas é zero, indicando dados de origem inconsistentes."
        )

    return _normalize_metric_keys(metrics)


def _generate_accounting_entries(documents: List[AuditedDocument]) -> List[AccountingEntry]:
    entries: List[AccountingEntry] = []
    for doc in documents:
        if doc.status == doc.status.ERRO or not doc.classification or not doc.doc.data:
            continue
        total_nfe = parse_safe_float(doc.doc.data[0].get("valor_total_nfe"))
        total_products = sum(parse_safe_float(item.get("produto_valor_total")) for item in doc.doc.data)
        total_icms = sum(parse_safe_float(item.get("produto_valor_icms")) for item in doc.doc.data)
        if total_nfe == 0 and total_products == 0:
            continue
        op_type = doc.classification.operationType
        if op_type == "Compra":
            entries.append(AccountingEntry(docName=doc.doc.name, account="1.1.2 Estoques", type="D", value=total_products))
            if total_icms > 0:
                entries.append(AccountingEntry(docName=doc.doc.name, account="1.2.1 ICMS a Recuperar", type="D", value=total_icms))
            entries.append(AccountingEntry(docName=doc.doc.name, account="2.1.1 Fornecedores", type="C", value=total_nfe))
        elif op_type == "Venda":
            entries.append(AccountingEntry(docName=doc.doc.name, account="1.1.3 Clientes", type="D", value=total_nfe))
            entries.append(AccountingEntry(docName=doc.doc.name, account="4.1.1 Receita de Vendas", type="C", value=total_products))
            if total_icms > 0:
                entries.append(AccountingEntry(docName=doc.doc.name, account="4.2.1 ICMS sobre Vendas", type="D", value=total_icms))
                entries.append(AccountingEntry(docName=doc.doc.name, account="2.1.2 ICMS a Recolher", type="C", value=total_icms))
        elif op_type == "Devolução":
            cfop = str(doc.doc.data[0].get("produto_cfop") or "")
            if cfop.startswith("1") or cfop.startswith("2"):
                entries.append(AccountingEntry(docName=doc.doc.name, account="2.1.1 Fornecedores", type="D", value=total_nfe))
                if total_icms > 0:
                    entries.append(AccountingEntry(docName=doc.doc.name, account="1.2.1 ICMS a Recuperar", type="C", value=total_icms))
                entries.append(AccountingEntry(docName=doc.doc.name, account="1.1.2 Estoques", type="C", value=total_products))
            else:
                entries.append(AccountingEntry(docName=doc.doc.name, account="4.1.2 Devoluções de Vendas", type="D", value=total_products))
                if total_icms > 0:
                    entries.append(AccountingEntry(docName=doc.doc.name, account="2.1.2 ICMS a Recolher", type="D", value=total_icms))
                entries.append(AccountingEntry(docName=doc.doc.name, account="1.1.3 Clientes", type="C", value=total_nfe))
        elif op_type == "Serviço":
            cfop = str(doc.doc.data[0].get("produto_cfop") or "")
            if cfop.startswith("5") or cfop.startswith("6") or cfop.startswith("7"):
                entries.append(AccountingEntry(docName=doc.doc.name, account="1.1.3 Clientes", type="D", value=total_nfe))
                entries.append(AccountingEntry(docName=doc.doc.name, account="4.1.3 Receita de Serviços", type="C", value=total_nfe))
            else:
                entries.append(AccountingEntry(docName=doc.doc.name, account="3.1.1 Despesa com Serviços", type="D", value=total_nfe))
                entries.append(AccountingEntry(docName=doc.doc.name, account="2.1.1 Fornecedores", type="C", value=total_nfe))
        elif op_type == "Transferência":
            cfop = str(doc.doc.data[0].get("produto_cfop") or "")
            if cfop.startswith("5") or cfop.startswith("6"):
                entries.append(AccountingEntry(docName=doc.doc.name, account="3.1.2 Custo de Transferência", type="D", value=total_products))
                entries.append(AccountingEntry(docName=doc.doc.name, account="1.1.2 Estoques", type="C", value=total_products))
            else:
                entries.append(AccountingEntry(docName=doc.doc.name, account="1.1.2 Estoques", type="D", value=total_products))
                entries.append(AccountingEntry(docName=doc.doc.name, account="4.1.4 Receita de Transferência", type="C", value=total_products))
    return entries


def _generate_sped(report: AuditReport) -> SpedFile:
    today = date.today()
    data_ini = date(today.year, today.month, 1).strftime("%d%m%Y")
    next_month = date(today.year + (today.month // 12), ((today.month % 12) + 1), 1)
    data_fim = date(next_month.year, next_month.month, 1) - date.resolution
    data_fim_str = data_fim.strftime("%d%m%Y")
    lines = [
        f"|0000|017|0|{data_ini}|{data_fim_str}|Nexus QuantumI2A2|12345678000195||SP|||A|1|",
        "|0001|0|",
        "|C001|0|",
    ]
    record_counts = {"0000": 1, "0001": 1, "C001": 1}

    def count(tag: str) -> None:
        record_counts[tag] = record_counts.get(tag, 0) + 1

    valid_docs = [doc for doc in report.documents if doc.status != doc.status.ERRO and doc.doc.data]
    for doc in valid_docs:
        first_item = doc.doc.data[0]
        lines.append(
            f"|C100|{'0' if doc.classification and doc.classification.operationType == 'Compra' else '1'}|0||55|||"
            f"{parse_safe_float(first_item.get('valor_total_nfe')):.2f}|"
            + "|" * 12
        )
        count("C100")
        aggregator: Dict[str, Dict[str, float]] = defaultdict(lambda: {"vBC": 0.0, "vIcms": 0.0, "vOper": 0.0})
        for item in doc.doc.data:
            cst = str(item.get("produto_cst_icms") or "00")[:3]
            cfop = str(item.get("produto_cfop") or "0000")
            aliq = f"{parse_safe_float(item.get('produto_aliquota_icms')):.2f}".replace(".", ",")
            key = f"{cst}|{cfop}|{aliq}"
            bucket = aggregator[key]
            bucket["vBC"] += parse_safe_float(item.get("produto_base_calculo_icms"))
            bucket["vIcms"] += parse_safe_float(item.get("produto_valor_icms"))
            bucket["vOper"] += parse_safe_float(item.get("produto_valor_total"))
        for key, values in aggregator.items():
            cst, cfop, aliq = key.split("|")
            lines.append(
                f"|C190|{cst}|{cfop}|{aliq}|{values['vOper']:.2f}|{values['vBC']:.2f}|{values['vIcms']:.2f}||||"
            )
            count("C190")
        for index, item in enumerate(doc.doc.data, start=1):
            lines.append(
                f"|C170|{index}|{item.get('produto_nome', '')}|{parse_safe_float(item.get('produto_qtd')):.2f}|UN|"
                f"{parse_safe_float(item.get('produto_valor_total')):.2f}||{item.get('produto_cfop')}|{item.get('produto_cst_icms')}||||"
            )
            count("C170")

    lines.append(f"|C990|{1 + record_counts.get('C100', 0) + record_counts.get('C170', 0) + record_counts.get('C190', 0)}|")
    lines.append("|9001|0|")
    lines.append("|0990|2|")
    total_lines = len(lines) + 2
    lines.append(f"|9990|{total_lines - 1}|")
    lines.append(f"|9999|{total_lines}|")
    return SpedFile(filename=f"SPED-EFD-{today.isoformat()}.txt", content="\n".join(lines))


def run_accounting_analysis(report: AuditReport) -> AuditReport:
    metrics = _run_deterministic_accounting(report)
    entries = _generate_accounting_entries(report.documents)
    sped = _generate_sped(report)

    valid_items = [item for doc in report.documents if doc.status != doc.status.ERRO and doc.doc.data for item in doc.doc.data]
    if not valid_items:
        summary = AnalysisResult(
            title="Análise Fiscal Concluída",
            summary="Não foram encontrados dados válidos para gerar um resumo detalhado.",
            keyMetrics=[{"metric": key, "value": str(value), "insight": ""} for key, value in metrics.items()],
            actionableInsights=[
                "Verificar a causa dos erros nos documentos importados para permitir uma análise completa."
            ],
            strategicRecommendations=[
                "Implementar controles de qualidade nos arquivos enviados ao pipeline fiscal."
            ],
        )
        return AuditReport(
            documents=report.documents,
            summary=summary,
            aggregatedMetrics=metrics,
            accountingEntries=entries,
            spedFile=sped,
            aiDrivenInsights=report.aiDrivenInsights,
            crossValidationResults=report.crossValidationResults,
            deterministicCrossValidation=report.deterministicCrossValidation,
        )

    key_metrics = [
        {"metric": key, "value": str(value), "insight": ""} for key, value in metrics.items() if key != "Alerta de Qualidade"
    ]
    actionable = [
        "Monitorar os produtos com maior variação de preço para prevenir distorções fiscais.",
        "Revisar CFOPs utilizados para garantir alinhamento com operações reais.",
    ]
    if "Alerta de Qualidade" in metrics:
        actionable.append(metrics["Alerta de Qualidade"])
    strategic = [
        "Aprimorar a governança de cadastros fiscais para reduzir riscos tributários.",
        "Avaliar oportunidades de crédito com base na estimativa de IVA calculada.",
    ]
    summary = AnalysisResult(
        title="Panorama Fiscal Consolidado",
        summary=(
            "A análise determinística consolidou os principais indicadores fiscais da base processada, permitindo decisões "
            "mais assertivas sobre compliance e planejamento tributário."
        ),
        keyMetrics=key_metrics,
        actionableInsights=actionable,
        strategicRecommendations=strategic,
    )

    return AuditReport(
        documents=report.documents,
        summary=summary,
        aggregatedMetrics=metrics,
        accountingEntries=entries,
        spedFile=sped,
        aiDrivenInsights=report.aiDrivenInsights,
        crossValidationResults=report.crossValidationResults,
        deterministicCrossValidation=report.deterministicCrossValidation,
    )
